mktestimg.c

```diff
@@ -22,87 +22,123 @@ enum {
-        if (!strcmp(argv[i], "-v")) {
-            printf("%s v%s\n", argv[0], MKTESTIMG_VERSION_STR);
-            return 0;
-        }
-        else if (!strcmp(argv[i], "-w") && width == 0) {
+    for (int i = 1; i < argc; i++)
+        if (!strcmp(argv[i], "-w") && width == 0)
             width = atoi(argv[++i]);
-        }
-        else if (!strcmp(argv[i], "-h") && height == 0) {
+
+        else if (!strcmp(argv[i], "-h") && height == 0)
             height = atoi(argv[++i]);
-        }
-        else if (!strcmp(argv[i], "-f") && format == 0) {
+
+        else if (!strcmp(argv[i], "-f") && format == 0)
             format = atoi(argv[++i]);
-        }
-        else if (!strcmp(argv[i], "-c") && channels == 0) {
+
+        else if (!strcmp(argv[i], "-c") && channels == 0)
             channels = atoi(argv[++i]);
-        }
+
         else {
-            if (!i) continue;
+            if (!i)
+                continue;
             if (!output) {
-                if (!strcmp(argv[i], "-o")) i++;
-                output = copystr(argv[i]);
-            }
-            else {
-                printf("unknown argument: %s\n", argv[i]);
-                usage(argv[0]);
-                return 0;
+                if (!strcmp(argv[i], "-o"))
+                    i++;
+                output = copy_str(argv[i]);
+            } else {
+                printf("Unexpected field: %s\n"
+                       "Try typing: %s - to get help.\n",
+                       argv[i], argv[0]);
+                return EXIT_FAILURE;
             }
         }
+
+    if (width == 0 || height == 0 || format == 0 || channels == 0 ||
+        output == 0) {
+        fprintf(stderr, "error: Arguments are malformed.\n");
+        return EXIT_FAILURE;
     }
 
     if (!(width && height)) {
-        fprintf(stderr, "%s: width or height cannot be zero\n", argv[0]);
-        return 0;
+        fprintf(stderr, "%s: width or height cannot be zero!\n", argv[0]);
+        return EXIT_FAILURE;
     }
 
-    if (!format || format >= FORMAT_COUNT) format = FORMAT_PNG;
-    if (!channels || channels > 4 || channels < 3) channels = 4;
-    if (!output) output = copystr("out.img");
+    if (!format || format >= FORMAT_COUNT)
+        format = FORMAT_PNG;
+    if (!channels || channels > 4 || channels < 3)
+        channels = 4;
+    if (!output)
+        output = copy_str("out.jpg");
 
-    printf("width: %d\nheight: %d\nchannels: %d\noutput: %s\n", width, height, channels, output);
+    printf("width: %d\n"
+           "height: %d\n"
+           "channels: %d\n"
+           "output: %s\n",
+           width, height, channels, output);
 
     stbi_write_png_compression_level = 1;
-    char *data = malloc(sizeof(char) * width * height * channels);
-    
-    for (int i = 0; i < width * height; i++) {
-        data[i * channels + 0] = 255;
+    char* data = malloc(width * height * channels);
+
+    for (int i = 0, res = width * height; i < res; i++) {
```

Comment:
May i know the reason for creating another temp variable to store width * height?
Well, if you're concerned about performance as in width * height is gonna recomputed each time the loop is run then it doesn't work like that. Most compilers do enough optimization to not let that happen. Even if some compiler doesn't optimizes it, the program itself is a very short lived one, so such optimizations wouldn't matter. In fact removing free() is also accepted (D programming language does) because OS automatically frees all the unfreed memory.

Reply:
_because OS automatically frees all the unfreed memory._ – Then the programmers shouldn't have used `free()` ever. Don't you think it's a good practice to minimize the comparison statement rather than putting an expression there? That's what I've done.

Reply:
> Don't you think it's a good practice to minimize the comparison statement rather than putting an expression there?
I generated the assembly output, and what you are saying is right :+1: thanks

> Then the programmers shouldn't have used free() ever.
It was just a side note but there are actual professional programs which do not call free() because GPAs are slow and in short lived program, unfreed memory isn't that big of a problem. It **is** indeed a huge problem for long run programs though.